### PR TITLE
ES6: computed property names #26

### PR DESCRIPTION
**Pull request by ssrh**

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -1539,6 +1573,12 @@ export class Parser extends Tokenizer {
         return this.markLocation(new Shift.ShorthandProperty(new Shift.Identifier(key.value)), startTokenIndex);
       }
     }
+    if (this.match(TokenType.LBRACK)) {
```
Comment (ikarienator) [inline src/parser.js]: Maybe we can consider merging this test to the next if statement to avoid the duplicate code?

File: src/parser.js [hunk 2]
```diff
@@ -1489,6 +1496,33 @@ export class Parser extends Tokenizer {
     if (token instanceof IdentifierLikeToken) {
       return new Shift.StaticPropertyName(this.parseIdentifier().name);
     }
+    if (token.type == TokenType.LBRACK) {
+      let previousGeneratorParameter = this.paramGeneratorParameter;
+      let previousYield = this.paramYield;
+      let previousIn = this.paramIn;
```
Comment (ikarienator) [inline src/parser.js]: `paramIn` is not affected here.

File: src/parser.js [hunk 3]
```diff
@@ -1489,6 +1496,33 @@ export class Parser extends Tokenizer {
...
+      this.expect(TokenType.LBRACK);
+      if (this.paramGeneratorParameter) {
+        // [+GeneratorParameter] ComputedPropertyName
+        this.paramGeneratorParameter = false;
+        this.paramYield = false;
+        this.paramIn = true;
+      } else {
+        // [~GeneratorParameter] ComputedPropertyName[?Yield]
+        // this.paramYield as passed in
+        this.paramIn = true;
```
Comment (ikarienator) [inline src/parser.js]: This part is shared in both branches.

File: src/parser.js [hunk 4]
```diff
@@ -1489,6 +1496,33 @@ export class Parser extends Tokenizer {
...
+      }
+      let expr = this.parseAssignmentExpression();
+      token = this.lookahead;
+      if (token.type !== TokenType.RBRACK) {
```
Comment (ikarienator) [inline src/parser.js]: This is handled by `expect` already.

File: src/parser.js [hunk 5]
```diff
@@ -130,6 +130,11 @@ export class Parser extends Tokenizer {
     this.inIteration = false;
     this.inSwitch = false;
     this.inFunctionBody = false;
+    this.paramGeneratorParameter = false;
+    this.paramYield = false;
+    // 12.9 NOTE the [In] grammar parameter is needed to avoid confusing the in operator in a relational
+    //           expression with the in operator in a for statement
+    this.paramIn = false;
```
Comment (ikarienator) [inline src/parser.js]: `allowIn` needs to be renamed to `paramIn`, they are not separate params.

Comment (ssrh) [inline src/parser.js]: If allowIn is the same as paramIn, why is it initialized to true?

File: src/parser.js (hunk 1 above)
Comment (ssrh) [inline src/parser.js]: we can, but I need to understand why it is necessary to test the following before going back into this.parseObjectPropertyKey().  Couldn't we hit eof() at any point inside parseObjectPropertyKey?

 if (this.eof() || token.type.klass == TokenClass.Punctuator) {

Comment (ikarienator) [inline src/parser.js]: That's probably true. Maybe you can try to just remove that branch?

File: src/parser.js (hunk 5 above)
Comment (ikarienator) [inline src/parser.js]: Because Expression production is normally parsed with the `In` parameter.

**Comment by ikarienator** [comment]
Also, please rebase.

File: src/parser.js [hunk 6]
```diff
@@ -300,7 +306,10 @@ export class Parser extends Tokenizer {
         return this.markLocation(this.parseWithStatement(), startTokenIndex);
       default:
       {
+        let previousIn = this.paramIn;
```
Comment (ikarienator) [inline src/parser.js]: `allowIn = true` is the default. No need to set it here.

File: src/parser.js [hunk 7]
```diff
@@ -1489,6 +1595,25 @@ export class Parser extends Tokenizer {
     if (token instanceof IdentifierLikeToken) {
       return new Shift.StaticPropertyName(this.parseIdentifier().name);
     }
+    if (token.type == TokenType.LBRACK) {
+      let previousGeneratorParameter = this.paramGeneratorParameter;
+      let previousYield = this.paramYield;
+      let previousIn = this.paramIn;
```
Comment (ikarienator) [inline src/parser.js]: `paramIn` is always true at this point. No need to assigned it.

Comment (ssrh) [inline src/parser.js]: I believe that paramIn will be false when this code is reached via the last two choices of MethodDefinition[Yield], get and set

**Comment by michaelficarra** [comment]
I've rebased off `es6` branch for you. Please pull.

File: src/parser.js (hunk 7 above)
Comment (ikarienator) [inline src/parser.js]: That's not possible. When `MethodDefinition` is reached `paramIn` is always true.

File: test/object-expressions/property-name.js [hunk 8]
```diff
@@ -68,5 +68,32 @@ suite("Parser", function () {
...
+    testParse("({get [6+3]() {}, set [5/4](x) {}})", expr,
+      new Shift.ObjectExpression([
+        new Shift.Getter(new Shift.ComputedPropertyName(
+          new Shift.BinaryExpression("+", new Shift.LiteralNumericExpression(6), new Shift.LiteralNumericExpression(3))
+        ), new Shift.FunctionBody([], [])),
+        new Shift.Setter(new Shift.ComputedPropertyName(
+          new Shift.BinaryExpression("/", new Shift.LiteralNumericExpression(5), new Shift.LiteralNumericExpression(4))
+        ), new Shift.Identifier("x"), new Shift.FunctionBody([], [])),
+      ])
+    );
```
Comment (ikarienator) [inline test/object-expressions/property-name.js]: Can you add a failure test that `({[1,2]:3})` is not valid?

File: src/parser.js [hunk 9]
```diff
@@ -1638,6 +1642,26 @@ export class Parser extends Tokenizer {
...
+      let previousYield = this.paramYield;
+      this.expect(TokenType.LBRACK);
+      if (this.paramGeneratorParameter) {
+        // [+GeneratorParameter] ComputedPropertyName
+        this.paramGeneratorParameter = false;
+        this.paramYield = false;
+      } // else [~GeneratorParameter] ComputedPropertyName[?Yield]
+      let expr = this.parseAssignmentExpression();
+      token = this.lookahead;
+      if (token.type !== TokenType.RBRACK) {
```
Comment (ikarienator) [inline src/parser.js]: This if statement can be replaced by `this.expect(TokenType.RBRACK);`.

Comment (ssrh) [inline src/parser.js]: Won't expecting LBRAC change the error message from ErrorMessages.INVALID_PROPERTY_NAME to Expected '[', found ...

File: src/parser.js [hunk 10]
```diff
@@ -1693,14 +1713,12 @@ export class Parser extends Tokenizer {
...
-    } else {
-      let key = this.parseObjectPropertyKey();
-      this.expect(TokenType.COLON);
-      let value = this.parseAssignmentExpression();
-      return this.markLocation(new Shift.DataProperty(key, value), startTokenIndex);
-    }
+
+    // PropertyName[?Yield] : AssignmentExpression[In,?Yield]
+    let key = this.parseObjectPropertyKey();
+    this.expect(TokenType.COLON);
```
Comment (ikarienator) [inline src/parser.js]: This position allows either a `COLON` for `DataProperty` or a `LPAREN` for `Method`

File: test/object-expressions/property-name.js (hunk 8 above)
Comment (ikarienator) [inline test/object-expressions/property-name.js]: Please add a test for methods with `ComputedPropertyName`.

Comment (ssrh) [inline test/object-expressions/property-name.js]: current es6 branch fails for:
  3) Parser property name ({"moo"(x,y,z) {}}):
     [1:8]: Unexpected token (
  

  4) Parser property name ({3(x,y,z) {}}):
     [1:4]: Unexpected token (

Comment (ikarienator) [inline test/object-expressions/property-name.js]: They should both work. Let me ping Dwight.

**Comment by ikarienator** [comment]
Shall we close this issue?

**Comment by ikarienator** [comment]
@ssrh I'm closing it for now. If you need this one instead, please feel free to reopen it.

**Pull request closed without merging**